subsys/rtio/rtio_executor_simple.c

```diff
@@ -71,13 +73,26 @@ int rtio_simple_submit(struct rtio *r)
 	return 0;
 }
 
+static inline void rtio_simple_spsc_release(struct rtio *r, struct rtio_sqe *sqe)
+{
+	if (FIELD_GET(RTIO_SQE_MULTISHOT, sqe->flags) == 0) {
+		LOG_DBG("Release SQEs");
+		rtio_spsc_release(r->sq);
+	} else if (FIELD_GET(RTIO_SQE_MEMPOOL_BUFFER, sqe->flags) == 1 && sqe->op == RTIO_OP_RX) {
+		/* Multi-shot read with mempool buffer, reset the memory for the next read */
+		LOG_DBG("Multi-shot with mempool detected, clearing buffers for %p", (void *)sqe);
+		sqe->buf = NULL;
+		sqe->buf_len = 0;
```

Comment:
@teburd I'm having some difficulties with this one. So we have the flag for multishot and I want to allow other tasks to run which means really what I want to do is put this at the end of the `sq`, but... I can't really do that because then I'll have a new sqe handle for the `rtio_sqe_cancel()` operation. So I'm wondering if the solution is to make the `rtio_spsc_consume()` a bit more complicated by having the fallback (when nothing was found at HEAD) to scan the entire thing for valid SQEs. I don't like this solution, but I'm also not really seeing how we can avoid fragmentation with multishot support.

Reply:
I have a branch that resolves the issue with spsc head of line blocking here. It drops the rings (spsc queues) in favor of lists everywhere, so for getting a sqe, you pull from a free list, for pushing to the submission queue, its a mpsc_push, and so on.

I'll push it up later today, we can work out the ordering of the commits and such then.